### ISSUE TITLE
Impossible to scale docker-compose due to undefined variable 'project'

### ISSUE THREAD

**Issue by TeunRoefs** [puppetlabs/puppetlabs-docker#390]
## What you expected to happen?
Scaling the docker-compose to `2`
## What happened?
The module wasn't able to scale up the number of containers to the required amount due to an unspecified variable.

The variable was still defined in puppetlabs-docker 1.1.0: https://github.com/puppetlabs/puppetlabs-docker/blob/11dc4a66a86978f22ce057e5027fe20243d26916/lib/puppet/provider/docker_compose/ruby.rb#L81

It does no longer exist in puppetlabs-docker: 3.1.0 however, a reference to that variable still exists: https://github.com/puppetlabs/puppetlabs-docker/blob/4694e3d99a81ff28730039213e582311a0576ba2/lib/puppet/provider/docker_compose/ruby.rb#L71

Easy fix would be to change `project` to `name`

## How to reproduce it?
Create a .pp file and put the following code in:

```
if $docker_compose {
  $docker_compose.each |String $ymlfile, Hash $properties| {
    docker_compose { $ymlfile:
      compose_files => $properties[compose_files],
      ensure        => present,
      scale         => $properties[scale],
    }
  }
}
```
Create a hiera-definition as follows:
```
docker_compose:
  'centos-container':
    compose_files:
      - /root/docker/compose/centos-container/docker-compose.yml
    scale:
      'centos-dockercontainer': 2
```

Run puppet to see the error message.

## Anything else we need to know?
Tested the same docker-compose file without using `scale` and then the container is created. 

On another note, the container keeps being 'created' and it seems the module is not idempotent using docker-compose.

## Versions:
```
$ puppet --version
5.5.2
$ docker version
18.09.0
$ facter os
Ubuntu 18.04
$ puppet module list
```
├── puppetlabs-apt (v6.2.1)

├── puppetlabs-docker (v3.1.0)

├── puppetlabs-powershell (v2.2.0)

├── puppetlabs-reboot (v2.1.1)

├── puppetlabs-stdlib (v5.1.0)

└── puppetlabs-translate (v1.1.0)
## Logs:
```
Error: Could not set 'present' on ensure: undefined local variable or method `project' for Docker_compose[centos-container](provider=ruby):Puppet::Type::Docker_compose::ProviderRuby
Did you mean?  proc (file: /etc/puppetlabs/code/environments/xxx/modules/profile/manifests/linux/docker.pp, line: 36)
Error: Could not set 'present' on ensure: undefined local variable or method `project' for Docker_compose[centos-container](provider=ruby):Puppet::Type::Docker_compose::ProviderRuby
Did you mean?  proc (file: /etc/puppetlabs/code/environments/xxx/modules/profile/manifests/linux/docker.pp, line: 36)
Wrapped exception:
undefined local variable or method `project' for Docker_compose[centos-container](provider=ruby):Puppet::Type::Docker_compose::ProviderRuby
Did you mean?  proc
Error: /Stage[main]/Profile::Linux::Docker/Docker_compose[centos-container]/ensure: change from 'absent' to 'present' failed: Could not set 'present' on ensure: undefined local variable or method `project' for Docker_compose[centos-container](provider=ruby):Puppet::Type::Docker_compose::ProviderRuby
Did you mean?  proc (file: /etc/puppetlabs/code/environments/xxx/modules/profile/manifests/linux/docker.pp, line: 36)
```

Hopefully anyone can help make the module idempotent and fix this issue  as well.

Thank you in advance!

**Comment by TeunRoefs** [comment]
Update on the idempotency part, we had to rename our labels to match that of the `exists?` in the provider. It is working now, however, the scaling does not work because of the `project` variable

**Comment by TeunRoefs** [comment]
Seems the issue has already been fixed, just not a new tag on the module. https://github.com/puppetlabs/puppetlabs-docker/commit/059fe187ebc431bef9b5c3b21bba6b358ac99b8f

Could anyone create a new tag?

**Comment by davejrt** [comment]
Hi @TeunRoefs apologies for the delay. We are planning a release in the upcoming days that will contain this fix.

**Comment by TeunRoefs** [comment]
Hi @davejrt no worries, we forked the module for now and made some changes to fix the `exists?` as well. Scaling had some issues because of the `compose_containers.uniq!`. We've fixed it using the following code.

```
def exists?
    Puppet.info("Checking for compose project #{name}")
    compose_services = {}
    compose_containers = []
    total_compose_containers = 0
    resource[:compose_files].each do |file|
      compose_file = YAML.safe_load(File.read(file), [], [], true)
      containers = docker([
                            'ps',
                            '--format',
                            "{{.Label \"com.docker.compose.service\"}}-{{.Image}}",
                            '--filter',
                            "label=com.docker.compose.project=#{name}",
                          ]).split("\n")
      compose_containers.push(*containers)
      total_compose_containers += compose_containers.count
      compose_containers.uniq!
      case compose_file['version']
      when %r{^2(\.[0-3])?$}, %r{^3(\.[0-6])?$}
        compose_services.deep_merge!(compose_file['services'])
      # in compose v1 "version" parameter is not specified
      when nil
        compose_services.deep_merge!(compose_file)
      else
        raise(Puppet::Error, "Unsupported docker compose file syntax version \"#{compose_file['version']}\"!")
      end
    end
 
    if compose_services.count != compose_containers.count
      return false
    end
 
    counts = Hash[*compose_services.each.map { |key, array|
                    image = (array['image']) ? array['image'] : get_image(key, compose_services)
                    Puppet.info("Checking for compose service #{key}")
                    ["#{key}", total_compose_containers]
                  }.flatten]
 
    # No containers found for the project
    if counts.empty? ||
       # Containers described in the compose file are not running
       counts.any? { |_k, v| v.zero? } ||
       # The scaling factors in the resource do not match the number of running containers
       resource[:scale] && counts.merge(resource[:scale]) != counts
      false
    else
      true
    end
  end
```

**Comment by davejrt** [comment]
Fixed in https://github.com/puppetlabs/puppetlabs-docker/commit/059fe187ebc431bef9b5c3b21bba6b358ac99b8f which was released as part of 3.2.0
